lib/HLSL/HLExpandStoreIntrinsics.cpp

```diff
@@ -0,0 +1,136 @@
+// License. See LICENSE.TXT for details.                                     //
+//                                                                           //
+///////////////////////////////////////////////////////////////////////////////
+
+#include "dxc/Support/Global.h"
+#include "dxc/HLSL/HLOperations.h"
+#include "dxc/HlslIntrinsicOp.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Intrinsics.h"
+#include "llvm/Transforms/Scalar.h"
+
+using namespace hlsl;
+using namespace llvm;
+
+namespace {
+
+// Expands buffer stores of aggregate value types
+// into stores of its individual elements, 
+// before SROA happens and we lose the layout information.
+class HLExpandStoreIntrinsics : public FunctionPass {
+public:
+  static char ID;
+  explicit HLExpandStoreIntrinsics() : FunctionPass(ID) {}
+
+  const char *getPassName() const override {
+    return "Expand HLSL store intrinsics";
+  }
+
+  bool runOnFunction(Function& Func) override;
+
+private:
+  bool expand(CallInst *StoreCall);
+  void emitElementStores(CallInst &OriginalCall,
+    SmallVectorImpl<Value*>& GEPIndicesStack, Type *StackTopTy,
+    unsigned OffsetFromBase);
+};
+
+char HLExpandStoreIntrinsics::ID = 0;
+
+bool HLExpandStoreIntrinsics::runOnFunction(Function& Func) {
+  bool changed = false;
+
+  for (auto BlockIt = Func.begin(); BlockIt != Func.end(); BlockIt++) {
+    BasicBlock &Block = *BlockIt;
+    for (auto InstIt = Block.begin(); InstIt != Block.end();) {
+      CallInst *Call = dyn_cast<CallInst>(InstIt++);
+      if (Call == nullptr
+        || GetHLOpcodeGroup(Call->getCalledFunction()) != HLOpcodeGroup::HLIntrinsic
+        || static_cast<IntrinsicOp>(GetHLOpcode(Call)) != IntrinsicOp::MOP_Store) {
+        continue;
+      }
+
+      changed |= expand(Call);
+    }
+  }
+  return changed;
+}
+
+bool HLExpandStoreIntrinsics::expand(CallInst* StoreCall) {
+  Value *OldStoreValueArg = StoreCall->getArgOperand(HLOperandIndex::kStoreValOpIdx);
+  Type *OldStoreValueArgTy = OldStoreValueArg->getType();
+  // Only expand if the value argument is by pointer, which means it's an aggregate.
+  if (!OldStoreValueArgTy->isPointerTy()) return false;
+  
+  IRBuilder<> Builder(StoreCall);
+  SmallVector<Value*, 4> GEPIndicesStack;
+  GEPIndicesStack.emplace_back(Builder.getInt32(0));
+  emitElementStores(*StoreCall, GEPIndicesStack, OldStoreValueArgTy->getPointerElementType(), /* OffsetFromBase */ 0);
+  DXASSERT_NOMSG(StoreCall->use_empty()); // Should return void
+  StoreCall->eraseFromParent();
+  return true;
+}
+
+void HLExpandStoreIntrinsics::emitElementStores(CallInst &OriginalCall,
+    SmallVectorImpl<Value*>& GEPIndicesStack, Type *StackTopTy,
+    unsigned OffsetFromBase) {
+  llvm::Module &Module = *OriginalCall.getModule();
+  IRBuilder<> Builder(&OriginalCall);
+
+  if (StructType *StructTy = dyn_cast<StructType>(StackTopTy)) {
```

Comment:
how do you handle non-empty nested structs?

Reply:
This is a recursive function so it should work, did I make a mistake? I'll update the test.

Reply:
> This is a recursive function so it should work, did I make a mistake? I'll update the test.

never mind. I was expecting it to be recursive, but then totally overlooked that you do have a recursive implementation. and yeah, good to have a test or two to exercise this code a bit more..